src/jon_mikkel_ex/ex06/logistic_regression.py

```diff
@@ -0,0 +1,333 @@
+# -*- coding: utf-8 -*-
+
+__author__ = "Jon-Mikkel Korsvik"
+__email__ = "jonkors@nmbu.no"
+
+import numpy as np
+from numpy.linalg import norm
+from sklearn.base import BaseEstimator, ClassifierMixin
+from sklearn.exceptions import NotFittedError
+from sklearn.utils import check_random_state, check_X_y
+
+
+def sigmoid(z):
+    r"""Perform a logistic transform on the input.
+
+    This function applies the sigmoidal function element-wise to all
+    elements of `z`. The sigmoidal function is on the following form:
+
+    .. math::
+
+        \frac{1}{1 + exp(-\mathbf{z})}.
+
+    Parameters
+    ----------
+    z : np.ndarray
+        Logit to transform.
+
+    Returns
+    -------
+    sigmoidal_transformed_z : np.ndarray
+        Transformed input.
+    """
+    if type(z) is int or float:
+        return 1 / (1 + np.exp(-z))
+    if type(z) is np.ndarray:
+        sigmoidal_transformed_z = np.zeros_like(z, np.float64)
+        for ind, val in enumerate(z):
+            sigmoidal_transformed_z[ind] = 1 / (1 + np.exp(-val))
+        return sigmoidal_transformed_z
+    else:
+        raise TypeError('Z input has to be a np.ndarray, alternatively\
+ a float or an int will suffice')
+
+
+def predict_proba(coef, X):
+    r"""Predict the class probabilities for each data point in :math:`X`.
+
+    Estimate which class each data point in X corresponds to. This is done
+    according to the following formula.
+
+    .. math::
+
+        \hat{y}_i = \sigma(\mathbf{x}_i^T \mathbf{w}),
+
+    where :math:`x_i` is the i-th row in :math:`X` and :math:`\sigma` is
+    the sigmoidal function. Alternatively, in matrix-vector form:
+
+    .. math::
+
+        \hat{\mathbf{y}} = \sigma(X \mathbf{w}).
+
+    Parameters
+    ----------
+    coef : np.ndarray(shape=(r,))
+        The weight vector, :math:`w`
+    X : np.ndarray(shape=(n, r))
+        The data matrix (aka design or measurement matrix)
+
+    Returns
+    -------
+    p : np.ndarray(shape(n,))
+        The predicted class probabilities.
+    """
+
+    return sigmoid(X@coef)
+
+
+def logistic_gradient(coef, X, y):
+    r"""Returns the gradient of a logistic regression model.
+
+    The gradient is given by
+
+    Parameters
+    ----------
+    coef : np.ndarray(shape=(r,))
+        The weight vector, :math:`w`
+    X : np.ndarray(shape=(n, r))
+        The data matrix (aka design or measurement matrix)
+    y : np.ndarray(shape=(n,))
+        The true class labels for each data point.
+
+    Returns
+    -------
+    gradient : np.ndarray(shape=(r,))
+        The gradient of the cross entropy loss related to the linear
+        logistic regression model.
+    """
+    proba = predict_proba(coef, X)
+    return X.T@(y - proba)
```

Comment:
if you experienced a bad accuracy while running the `__main__` section, this might be why. According to https://inf200-coursework.readthedocs.io/en/latest/, it should be `(y_hat - y)` not the opposite (as was written in the script). But it looks like you have accounted for it in line 228, so it should still converge.